File: ios/gk-ios/Utils/MiniPlayerSlot.swift
import Foundation

/// Pure state resolution for the mini-player slot above the tab bar (docs/screens/player.md **v15**
/// — "The mini-player is never empty").
///
/// **Why this is a separate, `Foundation`-only file.** The whole point of v15 is a *precedence*
/// rule ("whatever is loaded in the player wins; the last-visited song only fills the slot when the
/// player is empty"), and precedence is the one part of the bar that can be verified without a
/// device or a running `AVPlayer`. Kept beside `TakeQueue` for the same reason and in the same
/// shape: no SwiftUI, no AVFoundation, every branch reachable from
/// `gk-iosTests/MiniPlayerSlotTests.swift`.
///
/// Names are kept identical across platforms (`MiniPlayerSlot`, `resolveMiniPlayerSlot`) so the
/// same concept reads the same way on Android (CLAUDE.md "Parallel structure & naming").

/// What the mini-player slot is showing right now (player.md v15 state table).
///
/// The three cases are exhaustive and mutually exclusive, so `MiniPlayerView` can `switch` over
/// them and cannot accidentally render two bars — or, as before v15, none.
enum MiniPlayerSlot: Equatable {
    /// **The track state.** A take is loaded — playing, paused, loading *or* errored — so the bar
    /// shows the loaded song, its reciter, and a play/pause transport.
    case playing(song: Song)

    /// **The resting state.** Nothing is loaded, but this song was the last one the reader opened.
    /// The bar keeps its size and position and shows the song's **author** (no take is chosen, so
    /// there is no reciter to name — see docs/screens/tracks.md "reciter vs author").
    ///
    /// - Parameter canPlay: `audio_available` for that song. `true` gives the single play
    ///   affordance that starts its first take; `false` keeps the slot but replaces that affordance
    ///   with an open-song chevron ("the reader was there, it is still the way back to it").
    case resting(song: ManifestEntry, canPlay: Bool)

    /// Nothing to show at all — a fresh install where no song has ever been opened. The only state
    /// in which the slot takes no vertical space.
    case absent
}

/// Resolves which state the mini-player slot is in, given the player and the persisted
/// last-visited song (player.md v15).
///
/// Precedence, in order:
///
/// 1. **A loaded song always wins.** It is what the reader is listening to; the resting state must
///    never displace it.
/// 2. Otherwise the **last visited song**, rehydrated from the bundled corpus — the resting state
///    "only fills the gap".
/// 3. Otherwise **absent**.
///
/// - Parameters:
///   - currentSong: `AudioPlayerService.currentSong` — non-`nil` from the moment a song is loaded,
///     **including** a play attempt that failed. `play(song:track:)` assigns it before its
///     "nothing to play" guard fires precisely so that invariant holds; without it an errored song
///     with no takes would fall through to the resting state and swap the bar's song out from
///     under the reader.
///   - state: the player's state machine. It deliberately does **not** change the outcome: a
///     loaded song holds the slot in *every* state, `.error` included, and the `currentSong`
///     invariant above is what guarantees `.error` is reachable here with a song in hand. (Android
///     spells the same rule as `nowPlaying != null || playbackState == ERROR`; that disjunction is
///     not expressible here, since `.playing` carries a non-optional `Song` — so iOS enforces it at
///     the source instead of patching it up in the resolver.) `state` is kept in the signature so
///     the call site states its intent and so the signature matches Android's — the same reason
///     `TakeQueue.resolveTakeEndAction` keeps its `shuffle` parameter.
///   - lastVisited: the `ManifestEntry` for the persisted `player.lastVisitedSongUid`, or `nil`
///     when nothing is persisted *or* the persisted uid is no longer in the corpus. Resolving it
///     to `nil` rather than trusting the stored uid is what makes a stale record degrade to an
///     absent slot instead of a crash (player.md v15 "Storage failures are swallowed").
func resolveMiniPlayerSlot(
    currentSong: Song?,
    state: AudioPlayerState,
    lastVisited: ManifestEntry?
) -> MiniPlayerSlot {
    if let currentSong {
        return .playing(song: currentSong)
    }
    if let lastVisited {
        return .resting(song: lastVisited, canPlay: lastVisited.audioAvailable)
    }
    return .absent
}

File: andorid/app/src/main/java/com/gaudiyakirtan/services/MiniPlayerSlot.kt
package com.gaudiyakirtan.services

/**
 * Pure resolution of **what the mini-player slot shows** (docs/screens/player.md v15, "The
 * mini-player is never empty").
 *
 * Kept **free of every Android import** -- no `Context`, no Compose, no `MediaPlayer` -- for the same
 * reason [resolveTakeEndAction] is (see `TakeQueue.kt`): the precedence rule is the only interesting
 * part of the change, and it belongs where a plain JVM JUnit test can reach it (see
 * `app/src/test/.../services/MiniPlayerSlotTest.kt`). iOS carries the same
 * `resolveMiniPlayerSlot` concept under the same name.
 *
 * The slot has exactly three shapes, matching the spec's table row for row.
 */

/**
 * The last song the reader opened, rehydrated from the bundled corpus (never from prefs -- only its
 * uid is persisted, see [com.gaudiyakirtan.data.LastVisitedRepository]).
 *
 * [author] is the song's **composer**, not a reciter: no take is chosen in the resting state, so
 * there is no reciter to name (docs/screens/player.md v15 "It is not a track"; docs/screens/tracks.md
 * "reciter vs author"). [audioAvailable] is what splits the bar's trailing affordance between a play
 * control and an open-song chevron.
 */
data class LastVisitedSong(
    val uid: String,
    val title: String,
    val author: String,
    val audioAvailable: Boolean
)

/** What the mini-player slot renders. One of these always holds outside song-detail. */
sealed interface MiniPlayerSlot {

    /**
     * The track state: a take is loaded -- playing, paused, loading or errored -- so the bar is the
     * player's, with the reciter credit and a play/pause transport. The loaded take is read from
     * [PlayerUiState] directly; this object only records that the player won the slot.
     */
    data object Playing : MiniPlayerSlot

    /**
     * The resting state: nothing is loaded, but [song] was opened at some point. Same bar, same
     * height, the composer's credit, and a single play affordance rather than a transport.
     */
    data class Resting(val song: LastVisitedSong) : MiniPlayerSlot

    /** A fresh install where no song has ever been opened -- there is genuinely nothing to resume. */
    data object Absent : MiniPlayerSlot
}

/**
 * The precedence rule, whole (docs/screens/player.md v15): **whatever is loaded in the player wins**;
 * the last-visited song only fills the slot when the player is empty.
 *
 * [playbackState] matters independently of [nowPlaying] because a song with no playable take leaves
 * the player in [PlaybackState.ERROR] with nothing loaded (see [PlayerViewModel.play]) -- that is
 * still the player occupying the slot with its "audio unavailable" message, and quietly replacing it
 * with the resting bar would swallow the failure the reader just caused.
 */
fun resolveMiniPlayerSlot(
    nowPlaying: NowPlaying?,
    playbackState: PlaybackState,
    lastVisited: LastVisitedSong?
): MiniPlayerSlot = when {
    nowPlaying != null || playbackState == PlaybackState.ERROR -> MiniPlayerSlot.Playing
    lastVisited != null -> MiniPlayerSlot.Resting(lastVisited)
    else -> MiniPlayerSlot.Absent
}
